ENH: `integrate.cubature`: support for infinite limits

ollybritton opened the pull request:
#### Reference issue
- #20252
- Builds on:
	- #21330
	- #21473
- Originally part of #21632, which is now being split into multiple PRs so it is easier to review.

#### What does this implement/fix?
This adds support to `cubature` for approximating integrals with one or more sets of infinite limits:

$$
\int^\infty_{-\infty} \int^\infty_{-\infty} e^{-x^2-y^2} \text dx \text dy
$$

```python
cubature(gaussian, [-np.inf, -np.inf], [np.inf, np.inf])
```

This is achieved by applying the same variable transformations used in `quad_vec` to each variable separately. Integrating over intervals of the form $[a, \infty)$, $(-\infty, b]$ is also supported.

This PR also adds supports for specifying problematic points in the integrand's domain through a `points` argument:

$$
\int^1_{-1} \frac{\sin x}{x} \text dx
$$

```python
cubature(sinc, [-1], [1], points=[[0]])
```

These additions maintain support for array API, with the exception of JAX, as the implementation of the variable transformations use in-place index assignment.

### Acknowledgements
Thanks to @izaid and @ev-br, who have provided lots of help implementing this functionality, and thanks to @lucascolley for help with array API support.

ollybritton committed:
ENH: add `points` for subdividing initial region

- Add `points` argument to `cubature`
- Break up the region used for the initial estimate so that all `points`
  lie on the edge of a subregion
- Add tests using the `points` argument

ollybritton committed:
ENH: support infinite limits via variable transformation

- Handle limits where a > b by flipping them around and keeping track of
  the sign
- Add `_VariableTransform` base class
- Add `_InfiniteLimitsTransform` class
- Apply infinite limits transformation if any of `a` or `b` are infinite
- Add test for limits where a > b
- Add test for infinite limits
- Add test to check that break points are correctly mapped under
  infinite limits transformation

ollybritton committed:
DOC: add details of `points` and infinite limits to docstring

- Update docstring to say that infinite limits are allowed in `a` and
  `b`
- Add `points` argument to Parameters section
- Describe the transformation used to handle infinite limits in the
  docstring
- Add example of 2D integral with infinite limits
- Add example of 1D integral with singularities avoided using `points`

ollybritton committed:
TST: call `cubature` using keyword arguments in new tests

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -114,6 +114,12 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
         evaluating the population in parallel. This evaluation is carried out as
         ``workers(func, iterable)``.
+    points : list of array_like, optional
+        List of points by which to split up the initial region of integration. If the
+        rule being used does not evaluate `f` on the boundary of a region (which is the
+        case for all Genz-Malik and Gauss-Kronrod rules) then `f` will not be evaluated
+        at these points. This should be a list of array-likes where each element has
+        length ``ndim``. Default is empty. See Examples.
```
The example is useful but I think an extra sentence here too wouldn't hurt explaining why `points` not being evaluated here is useful

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -172,6 +178,24 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
+    integration variable will use the transformation :math:`x = b_i - \frac{1+t}{t}` and
+    :math:`t \in (0, 1)`.
+
+    In all three of these cases, the Jacobian of the transformation is
+    :math:`J(t) = t^{-2}`.
```
I think we could probably skip this detail, it is perhaps slightly confusing as there are n variables so the Jacobian is n by n, although I get the point it is making.

ollybritton replied:
Maybe I could rephrase this to "Jacobian determinant"? I think this would make more sense, especially since the details of the transformations above are for just one variable at a time.

j-bowhay replied:
Sound good

ollybritton replied:
I've now removed this since the Jacobian determinant for the last transformation is $-t^{-2}$, not $t^{-2}$.

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -258,13 +282,46 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
 
     xp = array_namespace(a, b)
     max_subdivisions = float("inf") if max_subdivisions is None else max_subdivisions
+    points = [] if points is None else [xp.asarray(p, dtype=xp.float64) for p in points]
```
Is there a reason for not using the default floating type? or maybe something using `xp.result_type`

ollybritton replied:
Thanks for the review. I'm a bit confused here, would the "default floating type" be something like `dtype=xp.floating` or `dtype=float`? I'm applying `xp.asarray` to each element in the list so that `points` can be specified as `array_like`. If using `dtype=xp.float64` is a problem here, is it also a problem on the lines below this one?

```
# Convert a and b to arrays
a = xp.asarray(a, dtype=xp.float64)
b = xp.asarray(b, dtype=xp.float64)
```

j-bowhay replied:
Sorry this was a slightly rushed comment. What I meant was this forces everything to be `xp.float64` whereas in some libraries (pytorch) the default floating dtype is `xp.float32`. If we did something like `a = xp.asarray(a, dtype=xp.results_type(a, b, 1.0))` then we don't promote things unnecessarily. cc @lucascolley @mdhaber who can probably explain more coherently than me

mdhaber replied:
I think `_array_api.xp_broadcast_promote` with `force_floating=True` may be useful here.
https://github.com/scipy/scipy/blob/ef157bf12f15cbf47f04f2d38576499be5a7703b/scipy/_lib/_array_api.py#L587

j-bowhay replied:
It would also be good to add a test for different `dtype`s of `a` and `b` to check the `dtype` of the result is promoted correctly

ollybritton replied:
Just writing some tests for this and found the following case: Rules like `gk21` return the nodes and weights as arrays using `xp.asarray([0.996, ..., -0.996])` and so are using the default floating dtype in that array namespace.

Since the default floating dtype in Numpy is `float64`, this means that if a user does something like:

```
cubature(
        f,
        np.asarray([0], dtype=np.float32),
        np.asarray([1], dtype=np.float32),
)
```

then the result will actually have dtype `float64` rather than `float32` because calculations are done with the nodes and weights, which themselves will have dtype `float64`. If they were using `torch` instead, then the result would have dtype `float32`. Is this what should happen?

If the nodes and weights should also be returned with dtype `float32` in the NumPy case, then I think it would be necessary to add a new parameter like `result_dtype` to the constructor of each `FixedRule`, since each `Rule` can be initialised independently without `a` or `b`.

mdhaber replied:
> Is this what should happen?

I think the whole calculation should respect the common dtype of the user input. 

> Rules like gk21 return the nodes and weights as arrays using `xp.asarray([0.996, ..., -0.996])` and so are using the default floating dtype in that array namespace.

It sounds like this would mean that even if the user were to provide `float64` input, the _actual_ precision of the output is no better than the default dtype of the array library. Is that right?

I'd suggest storing the nodes and weights with `float64` and downcasting if necessary before use (this is what I do, anyway) or adding a `dtype` parameter.

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -298,10 +355,45 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
+    #
+    # This means ``f`` won't be evaluated there if the rule being used has no evaluation
+    # points on the boundary.
+    if points == []:
```
prehaps more idiomatic?
```suggestion
    if len(points) == 0:
```

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+
+
+def _is_strictly_in_region(point, a, b):
+    xp = array_namespace(point, a, b)
```
Can we add `xp` as an argument to skip the `array_namespace` call?

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+    any of the subregions.
+    """
+
+    xp = array_namespace(a, b)
```
same here

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+    return regions
+
+
+class _VariableTransform:
```
Are there other transforms planned?

ollybritton replied:
Yes, integrals with functions as limits are also calculated using a transform. But if you think two transforms aren't enough to justify the base class, then I'm happy to remove this.

j-bowhay replied:
If there's stuff to come then I think this is fine

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+    """
+
+    def __init__(self, f, a, b):
+        self._xp = array_namespace(a, b)
```
Consider passing `xp` as an argument

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+        a = xp_copy(self._orig_a)
+        b = xp_copy(self._orig_b)
+
+        for index in self._double_inf_pos:
+            a[index] = -1
+            b[index] = 1
+
+        for index in self._semi_inf_pos:
+            a[index] = 0
+            b[index] = 1
```
```suggestion
        a[self._double_inf_pos] = -1
        b[self._double_inf_pos] = 1

        a[self._semi_inf_pos] = 0
        b[self._semi_inf_pos] = 1
```

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+    def inv(self, x):
+        t = xp_copy(x)
+
+        for i in self._negate_pos:
+            t[..., i] *= -1
+
+        for i in self._double_inf_pos:
+            t[..., i] = 1/(x[..., i] + self._xp.sign(x[..., i]))
+
+        for i in self._semi_inf_pos:
+            t[..., i] = 1/(x[..., i] - self._orig_a[i] + 1)
```
Again these are arrays right so we can skip the loop and use a list of indexes

ollybritton replied:
I've tried this but am getting errors using `array_api_strict`. Switching to boolean indexing also doesn't work since `t[..., self._semi_inf_pos]` raises errors about not being able to use boolean indexing and `...` at the same time.

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +469,226 @@ def _process_subregion(data):
+        x = xp_copy(t)
+        jacobian = 1.0
+
+        for i in self._negate_pos:
+            x[..., i] *= -1
+
+        for i in self._double_inf_pos:
+            # For (-oo, oo) -> (-1, 1), use the transformation x = (1-|t|)/t.
+            x[..., i] = (1 - self._xp.abs(t[..., i])) / t[..., i]
+            jacobian *= 1/(t[..., i] ** 2)
+
+        for i in self._semi_inf_pos:
+            # For (start, oo) -> (0, 1), use the transformation x = start + (1-t)/t.
+            x[..., i] = self._orig_a[i] + (1 - t[..., i]) / t[..., i]
+            jacobian *= 1/(t[..., i] ** 2)
```
same here re indexing

j-bowhay commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -388,6 +475,26 @@ def test_zero_width_limits(self, xp):
+        xp_assert_close(
+            res.estimate,
+            -basic_1d_integrand_exact(n, xp),
+            rtol=1e-1,
```
This is a very loose tol, any reason why?

j-bowhay reviewed (changes requested):
Thanks @ollybritton a first pass over things

ollybritton committed:
DOC: add more detail to `points` parameter

ollybritton committed:
DOC: rephrase to "Jacobian determinant" rather than "Jacobian"

fancidev reviewed (commented):
Good stuff! Just some cosmetic editorial comments.

ollybritton committed:
MAINT: pass `xp` as argument rather than calling `array_namespace`

ollybritton commented on scipy/integrate/_cubature.py:
```diff
@@ -172,6 +178,24 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
+    :math:`t \in (0, 1)`.
+
+    If :math:`a_i` and :math:`b_i` range over :math:`x \in (-\infty, b_i]`, the i-th
+    integration variable will use the transformation :math:`x = b_i - \frac{1+t}{t}` and
```
I think $b_i - (1+t)/t$ is correct here, since the $(-\infty, b_i]$ case is handled by flipping the limits, negating $x$ and then applying the transformation for $[a_i, \infty)$.

fancidev replied:
But if $0 < t < 1$ and $x = b_i - (1+t)/t$, then $x$ will range from $-\infty$ to $b_i-2$, and there is a “gap” $[b_i-2,b_i]$, right?

ollybritton replied:
Ah yes, thanks!

ollybritton committed:
MAINT: return results in the common dtype of input

- Store nodes and weights by default using `float64`
- Downcast if necessary to match the common floating dtype of user input
- Add tests for this

ollybritton committed:
MAINT: pass `xp` to `_apply_fixed_rule` rather than using `array_namespace`

ollybritton committed:
STY: `len(...) == 0` rather than comparing to `[]`

ollybritton committed:
TST: tighten tolerance on some tests

Tighten tolerances from 1e-1 to 1e-8 on the following tests:

- `test_pass_str`
- `test_pass_array_like_not_array`
- `test_zero_width_limits`
- `test_limits_other_way_around`
- `test_base_1d_quadratures_simple`
- `test_base_1d_quadratures_error_from_difference`

ollybritton committed:
DOC: correct details of transformation used for infinite limits

- Correct details of transformation used in the (\infty, b_i] case
- Remove details of Jacobian determinant since it is not true when the
  limits are (\infty, b_i]

ollybritton committed:
STY: remove extra blank line

ollybritton committed:
MAINT: remove unnecessary `negate_pos` in transformation

- `negate_pos` wasn't having any effect since the values that were
  negated would be overriden below

ollybritton committed:
DOC: correct details for (-\infty, b_i]

ollybritton commented:
I've just pushed some changes hopefully addressing most of what is above:

- Fix the details of the transformations being used in the docstring
- Pass `xp` to `_split_subregion`, `_is_strictly_in_region` and `_apply_fixed_rule` rather than repeatedly calling `array_namespace`
- Respect the dtype of the inputs when returning results
- Tighten tolerances on some tests

Also, I realised that `_negate_pos` in `_InfiniteLimitsTransform` code was redundant, so I've removed it. Also from above:

- Indexing arrays with a list rather than looping over indices: I'm getting errors using `array_api_strict` for this, and using boolean indexing also isn't working in expressions like `x[..., self._semi_inf_pos]` where `_semi_inf_pos` is a boolean array.
- Nodes and weights are now float64 by default rather than the default floating dtype of the array library.

lucascolley commented:
> using boolean indexing also isn't working in expressions like x[..., self._semi_inf_pos] where _semi_inf_pos is a boolean array.

https://data-apis.org/array-api/draft/API_specification/indexing.html#boolean-array-indexing seems to suggest that, for example, you can replace:

```python
In [13]: x.shape
Out[13]: (2, 2)
...
In [10]: x[..., np.asarray([False, True])]
Out[10]: 
array([[2],
       [4]])
```

with

```python
In [20]: np.expand_dims(x[np.asarray([[False, True], [False, True]])], axis=1)
Out[20]: 
array([[2],
       [4]])
```

I'm not sure how this generalises, but maybe you can do something like that.

ollybritton commented:
@lucascolley I've just run a very informal comparison on my machine, but it seems like using a for loop is comparable in performance to using boolean arrays in NumPy:

---

- x shape: `(10, 10, 10)`
- ix: `[2, 4, 6]`

|Method|Time|
|------|----|
|`[..., ix]`|8.32 μs ± 90.5 ns|
|`[..., ix_bool]`|9.95 μs ± 51.7 ns|
|for loop: 8.05 μs|89.7 ns|

---

- Similar to what will actually occur with a 5D integrand
- x shape: `(21**5, 5)`
- ix: `[1, 2]`

|Method|Time|
|------|----|
|`[..., ix]`|40.9 ms ± 1.87 ms|
|`[..., ix_bool]`|38.9 ms ± 1.36 ms|
|for loop|13.3 ms ± 203 μs|

---

The notebook I used to check:

```python
data = np.random.rand(21**5, 5)
ix = [1, 2]
ix_bool = [False, True, True, False, False]
```

```python
%%timeit
data[..., ix] *= 2
data[..., ix] /= 2
```

```python
%%timeit
data[..., ix_bool] *= 2
data[..., ix_bool] /= 2
```

```python
%%timeit
for i in ix:
    data[..., i] *= 2
    data[..., i] /= 2
```

Is it worth trying the other method or do you think a for loop over the indicies is acceptable?

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +471,211 @@ def _process_subregion(data):
+        self._semi_inf_pos = []
+        self._double_inf_pos = []
+
+        ndim = xp_size(a)
```
Currently unused
```suggestion
```

j-bowhay replied:
Oops sorry I confused myself as i was making local edits

j-bowhay committed:
remove for loop from __init__

j-bowhay committed:
still bugs to figure out

j-bowhay committed:
Update scipy/integrate/_cubature.py

Co-authored-by: Olly Britton <ollybritton@gmail.com>

j-bowhay committed:
fix axis location

j-bowhay committed:
fix summing boleans

ollybritton committed:
don't use arr[..., bool_array]

ollybritton committed:
Merge branch 'j-bowhay-remove_for_loops' into infinite-limits-and-points

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -298,10 +357,45 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
 
             rule = ProductNestedFixed([base_rule] * ndim)
```
Sorry not directly this PR but it would be great to add test coverage to line 356

ollybritton replied:
Great, I'll make a note of this.

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +471,233 @@ def _process_subregion(data):
+            #
+            # This case occurs when points are given by the user to avoid, but after
+            # applying a transformation, they are removed.
+            continue
```
Again not blocking but would it be possible to add a test which hits this line?

ollybritton replied:
Sorry, should have bumped this earlier. There's now a test which hits this in the latest commit.

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -374,3 +471,233 @@ def _process_subregion(data):
+    def points(self):
+        # If there are infinite limits, then the origin becomes a problematic point
+        # due to a division by zero there.
+        if self._num_inf != 0:
```
According to the test coverage this condition is always true

j-bowhay replied:
I think because of the way it is used by `cubature`

ev-br replied:
FWIW, I'd suggest keeping the code as is then, and add a comment. The less footguns for future development, the better.

j-bowhay reviewed (commented):
Some minor points but otherwise in good shape

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -872,6 +1260,45 @@ def test_genz_malik_1d_raises_error(self, xp):
+            lambda x: f_with_problematic_points(x, points, xp),
+            xp.asarray(a, dtype=xp.float64),
+            xp.asarray(b, dtype=xp.float64),
+            xp=xp,
```
@ollybritton here's the culprit for the array API failure. Either `xp` needs to come from `array_namespace` here, or `_InfiniteLimitsTransform` needs to apply `array_namespace` to the input `xp` during setup.

ollybritton replied:
Thanks!

ollybritton committed:
BUG: use compatible namespace in test for infinite limits and points

ollybritton committed:
BUG: correctly map break points when using a transformation

- Need to reshape points so that they can be mapped by `f.inv`
- Add test case for 1D and 2D integrals where the points specified are
  removed by the transformation
- Adjust double infinite transformation to prevent division by zero

j-bowhay commented:
@ollybritton, just to let you from tomorrow onwards I'll be away travelling for the next four months without much computer access. I might be able to have a look on the mobile app but hopefully someone else will be able to pickup the review for your final pr for arbitrary regions.

ev-br commented:
Where does this PR stand w.r.t. review comments?

j-bowhay commented:
> Where does this PR stand w.r.t. review comments?

I think it is just https://github.com/scipy/scipy/pull/21657#discussion_r1790529005 that needs resolving

ollybritton committed:
MAINT: explain check for any infinite limits in `points`

[lint only]

j-bowhay reviewed (approved):
LGTM provided CI is green!

ev-br commented:
Congrats @ollybritton ! Do we have a tracking issue for possible follow-ups to the `cubature` work? If not, could you please open one.

lucascolley commented:
Yeah, I'd suggest closing gh-20252 with a summary of what has been implemented, and opening a new issue with further ideas.
